### PR TITLE
 [FLINK-17258][tests] Enabling unaligned checkpoint in all tests by default

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

To harden unaligned checkpoints as quickly as possible, we want to enable it by default for all tests similarly how we did it for credit-based flow control network stack.

## Brief change log

- Checking system property "execution.checkpointing.unaligned" if no explicit configuration option has been used.
- Setting system property "execution.checkpointing.unaligned"="true" for maven surefire for all unit and IT tests.
- Setting the configuration "execution.checkpointing.unaligned"="true" for all e2e tests.

## Verifying this change

Affects only tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by AHeise** [comment]
Based on https://github.com/apache/flink/pull/11814 , so ignore first commit. Last commit only temporary and helps to collect all failing cases.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java [hunk 1]
```diff
@@ -414,9 +417,19 @@ public void enableUnalignedCheckpoints() {
...
 
 	/**
-	 * Returns whether checkpoints should be persisted externally.
+	 * Disables unaligned checkpoints.
 	 *
-	 * @return <code>true</code> if checkpoints should be externalized.
+	 * @see #enableUnalignedCheckpoints()
+	 */
+	@PublicEvolving
+	public void disableUnalignedCheckpoints() {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java:425]: Do we really need this method? Isn't `enableUnalignedCheckpoints(false);` enough? `enableUnalignedCheckpoints()` also is kind of redundant but probably more common.

For another `boolean` property we have just a single setter `setPreferCheckpointForRecovery`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 2]
```diff
@@ -500,8 +500,9 @@ private void setVertexConfig(Integer vertexID, StreamConfig config,
 
 		config.setStateBackend(streamGraph.getStateBackend());
 		config.setCheckpointingEnabled(checkpointCfg.isCheckpointingEnabled());
-		config.setUnalignedCheckpointsEnabled(checkpointCfg.isUnalignedCheckpointsEnabled());
 		config.setCheckpointMode(getCheckpointingMode(checkpointCfg));
+		config.setUnalignedCheckpointsEnabled(config.getCheckpointMode() == CheckpointingMode.EXACTLY_ONCE &&
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:504]: Isn't this an independent bug fix?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 3]
```diff
@@ -958,10 +960,10 @@ private void configureCheckpointing() {
 				.setMinPauseBetweenCheckpoints(cfg.getMinPauseBetweenCheckpoints())
 				.setMaxConcurrentCheckpoints(cfg.getMaxConcurrentCheckpoints())
 				.setCheckpointRetentionPolicy(retentionAfterTermination)
-				.setExactlyOnce(getCheckpointingMode(cfg) == CheckpointingMode.EXACTLY_ONCE)
+				.setExactlyOnce(exactlyOnce)
 				.setPreferCheckpointForRecovery(cfg.isPreferCheckpointForRecovery())
 				.setTolerableCheckpointFailureNumber(cfg.getTolerableCheckpointFailureNumber())
-				.setUnalignedCheckpointsEnabled(cfg.isUnalignedCheckpointsEnabled())
+				.setUnalignedCheckpointsEnabled(exactlyOnce && cfg.isUnalignedCheckpointsEnabled())
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:966]: This check seems duplicated `exactlyOnce && cfg.isUnalignedCheckpointsEnabled()` in two places. Maybe hide `exactlyOnce &&` check inside `isUnalignedCheckpointsEnabled`?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/OverWindowITCase.scala [hunk 4]
```diff
@@ -40,6 +40,11 @@ import scala.collection.mutable
 @RunWith(classOf[Parameterized])
 class OverWindowITCase(mode: StateBackendMode) extends StreamingWithStateTestBase(mode) {
 
+  @Before
+  def setupEnv(): Unit = {
+    env.getCheckpointConfig.disableUnalignedCheckpoints()
```
Comment (pnowojski) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/OverWindowITCase.scala:45]: Add a comment and link to a jira ticket why?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java (hunk 1 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java:425]: No we don't need it and I have no hard feelings for removing it. 

`disableUnalignedCheckpoints` would be relevant when UC becomes the default (see also `StreamExecutionEnvironment#disableOperatorChaining`).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java (hunk 2 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:504]: So far, we have assumed that the user only sets UC when checkpoint mode is exactly once. In this PR, we ignoring checkpoint mode and set it for all.

The question is where to address that:
- We could fail on that combination. Then we would need to disable UC for every at least once test. Not my favorite.
- We change checkpointing mode with a warning. Not my favorite.
- We disable/ignore UC (current way) with a warning (needs to be added).
- We could have a value `UC if exactly once` and make `checkpointCfg.isUnalignedCheckpointsEnabled()` more intelligent. Would also be plausible.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java (hunk 3 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:966]: Also possible way, it's pretty much ignoring UC when at least once is set. See also above for a deeper discussion.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java (hunk 1 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java:425]: So what about keeping and using just `setUnalignedCheckpoints(boolean)`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java (hunk 2 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:504]: I guess I would prefer the most:
> We could fail on that combination. Then we would need to disable UC for every at least once test. Not my favorite.

But this one is probably also fine:
> We disable/ignore UC (current way) with a warning (needs to be added).

But either way, I think this should deserve it's own ticket.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java (hunk 1 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java:425]: Yes, I'd probably go this way. Should we mend that also on 1.11?

Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java]: Extracted FLINK-18403, which will be backported (while this PR should only go to master).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java (hunk 2 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java]: Extracted FLINK-18403, which will be backported (while this PR should only go to master).

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/OverWindowITCase.scala (hunk 4 above)
Comment (AHeise) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/OverWindowITCase.scala]: Created two tickets (watermarks/rescale) and linked to them with a small comment.

**Comment by AHeise** [comment]
Thanks for the feedback @pnowojski . I forked out https://github.com/apache/flink/pull/12739 for API change and handling of inconsistent configuration. This PR is based on it, so that PR must be merged before this.

**Pull request merged**